core/src/main/java/org/apache/brooklyn/core/mgmt/internal/LocalEntityManager.java

```diff
@@ -212,13 +213,41 @@ public EntityTypeRegistry getEntityTypeRegistry() {
 
     @Override
     public Iterable<Entity> getAllEntitiesInApplication(Application application) {
+        // To fix https://issues.apache.org/jira/browse/BROOKLYN-352, we need to synchronize on
+        // preRegisteredEntitiesById and preManagedEntitiesById while iterating over them (because
+        // they are synchronizedMaps). entityProxiesById is a ConcurrentMap, so no need to 
+        // synchronize on that.
+        // Only synchronize on one at a time, to avoid the risk of deadlock.
+        
         Predicate<Entity> predicate = EntityPredicates.applicationIdEqualTo(application.getId());
-        Iterable<Entity> allentities = Iterables.concat(preRegisteredEntitiesById.values(), preManagedEntitiesById.values(), entityProxiesById.values());
-        Iterable<Entity> result = Iterables.filter(allentities, predicate);
-        return ImmutableSet.copyOf(Iterables.transform(result, new Function<Entity, Entity>() {
-            @Override public Entity apply(Entity input) {
-                return Entities.proxy(input);
-            }}));
+        Set<Entity> result = Sets.newLinkedHashSet();
+        
+        synchronized (preRegisteredEntitiesById) {
```

Comment:
Yeah, I wondered why it was protected. That would definitely not be my choice! I wouldn't even bother having a protected getter that provided direct access. Any such use-case would likely be making changes to core brooklyn, rather than a downstream project needing it - so we can add a getter if the need arises.